### DISCUSSION TITLE
Removing shading from dot matrix printouts

### DISCUSSION THREAD

**Discussion by sharygin** [ImageMagick/ImageMagick#7527] [Digital Image Processing] [answered]
I've got some scans of old scientific data from printouts off of a dot matrix printer that include incredibly challenging (for me) shading that I need to remove before I can OCR it. With some tips from posts on the old forums, I've been able to find some parameters that minimize it. The before (sample1) and after (out1) from my best efforts are shown here-- but my treatment is still producing more OCR errors than I expected. wondering if anyone has tips in particular on:
<li>how to improve the connected-components part, which is removing the dots as much as possible;
<li>how to improve the morphology part or an alternative, which is intended to damage the remaining dots along the edges of characters that are irreversibly merged with letters so that the conversion to b&w is more effective, or
<li>how to improve the thresholds when converting to b&w so that more of the parts softened by morphology are converted to white and not black

magick sample.png ( +clone -threshold 60% -negate -type bilevel -define connected-components:area-threshold=25 -define connected-components:mean-color=true -connected-components 4 ) -compose copy_opacity -composite -compose over -background white -flatten -deskew 40% -morphology close diamond -alpha off -threshold 50% out1.png

![sample1](https://github.com/user-attachments/assets/1ab1c6e9-1db9-4c56-94f1-4b744b2d8c03)
![out1](https://github.com/user-attachments/assets/35de8f3e-f649-4b3b-874d-a957f54f3d6f)

**Comment by fmw42** [comment]
You can try morphology. For example if on IM 7.x.x.x (please specify IM version and platform)

![img](https://github.com/user-attachments/assets/fb21235d-c54e-4def-8272-670ed9039a60)

`magick img.png -morphology close diamond:2 x.png`

![x](https://github.com/user-attachments/assets/9b81116e-90ef-4f29-a85a-1847e1973637)

**Reply by sharygin** [reply]
I'm on ImageMagick-7.1.1 on windows 10.
I've got a morphology step in there, but it's after connected components. alone, morphology does a pretty good job in your example-- however, it damages the numbers a lot by lightening the ends of the characters (for example the number 259 after Los Angeles, but many other 7s, 6,s 9s). And the ends of the characters are now the same gray as the faded dots, which makes it hard to filter them. Maybe there's a way to improve that? 

in the meantime, I'm also open to another approach, like a filter that can be run on the result, that performs a smoothing spline around the characters, that would trim the remaining 'burrs' off of the numbers and letters.

I think I've got a minor improvement, slightly less damaged from the dot removal, but it's not unambiguously better. 
magick img.png \
( +clone -threshold *50%* -negate -type bilevel -define connected-components:area-threshold=25 -define connected-components:mean-color=true -connected-components 4 ) -compose copy_opacity -composite -compose over -background white -flatten \ 
-morphology close diamond:1 \
-alpha off \
-*auto-threshold otsu* \
-deskew 40% out.png 

![out5](https://github.com/user-attachments/assets/bd4d5611-d0cc-4371-97a5-e4e668dab9d5)

**Reply by fmw42** [reply]
Looks pretty good to me.  Why is that not good enough?  Do you need to do OCR on it? Does OCR fail?

There is no one filter specifically for dot matrix removal in Imagemagick. You will have to script something as you have.

**Reply by sharygin** [reply]
Yeah I realize that I'm chasing diminishing returns. I have the feeling that the portion inside the parentheses could be dialed in better, I copied it from an old thread with a similar problem and I've tried a lot of variations on the parameters but I'm doing it blindly.

We will be processing several thousand images like the sample, and then OCR, so every 1% reduction in OCR errors from better image quality will help us a lot.

**Reply by sharygin** [reply]
To simplify down the ask given all the preceding -- is there a filter that would trim off the dot artifacts in this image? Does anyone understand the connected-components or filters well enough to recommend a way to clean up the noise here? 

![image](https://github.com/user-attachments/assets/08ccbb6a-10d2-45bb-bdaa-71e33f47dc61)

**Reply by fmw42** [reply, marked as answer]
At this scale, you can use

magick image.png -morphology close disk:3 result.png

![result](https://github.com/user-attachments/assets/dcb0880a-c458-4eaf-98ab-baeb692eac1f)

For smaller scale, you could try using disk:1 or smaller (as float) or go to diamond:X where x is an integer. See https://imagemagick.org/Usage/morphology/#kernel

**Reply by sharygin** [reply]
thanks, fmw42 -- I appreciate your suggestion to apply morphology again and will add this to my workflow, after the other processing. in my case disk:3 doesn't generalize to the entire page as I didn't include a 6 or 9 in this image, and those numbers' tails are damaged by the large kernel size, but disk/diamond:2 is working well.

**Reply by fmw42** [reply]
You can also open it up or erode the black letters after the close to make them thicker or fuller. Note that an erode on black makes it larger similar to dilating on white.

**Reply by sharygin** [reply]
Thanks for the suggestion. I attempted several variations on more morphology variants but didn't find anything that made an overall improvement. 

The sample is just one page of many. I have about 20 batches, each with about 300 pages, and sometimes the scan characteristics are different between batches. For each batch, I'm trying tweaks to the first +clone -threshold XX%, and the final -morphology close XX:X commands and balancing tradeoffs between noise and legibility for each set of images. 

I wish I could selectively identify parts of the image with dots and treat only those, leaving the areas without shading alone, because what's good treatment for the shaded areas is usually bad for the regular areas.

**Reply by fmw42** [reply]
You could create a mask based upon the average brightness of each row of the input (just average the image to one column and look at the brightness using txt: output to find a threshold value. You can use morphology if needed to fill in either before or afterwards. Then use the mask with your processing to process only the darker areas of the input.  The process the whole image. Then use

`magick input processed mask -compose over -composite result`

This will only keep the text from being processed in alternating bands.

**Reply by fmw42** [reply]
You are probably doing this already. But you could threshold to remove the dots as they seem slightly brighter than the text. Then perhaps erode on black to enlarge the text if the threshold removed some of the antialiasing at the edges of the characters. Then use that as a mask to process the whole image to remove the dots.

`magick input processed mask -compose over -composite result`

**Reply by sharygin** [reply]
When we started this project, we hoped there was a brightness difference between the dots and text that we could exploit with a low pass filter. We tried to plan ahead by testing different scan settings, but in the end the scans we got seemed to have more intensity and variability than we expected. It took that team many hours to scan and we can't ask them to try again unfortunately. 

I've tried but I'm coming up short trying to realize your advice about processing bands and masks. However in the process of reading your comments I realized that the workflow might benefit from an earlier morphology done to lighten the dots before the first threshold. That's working well. The tradeoffs are balanced very well between noise/clipping across many sample pages we've looked at. Compared to the flawed originals, I couldn't ask for much more. 

for %a in (*.png) DO magick "%~a" -morphology close diamond:2 ( +clone -auto-threshold otsu -negate -type bilevel -define connected-components:area-threshold=25 -define connected-components:mean-color=true -connected-components 4 ) -compose copy_opacity -composite -compose over -background white -flatten -morphology close diamond:2 -deskew 40% -alpha off -set colorspace Gray -colors 256 "%~dpna_clean_v5.png"

Compared to the previous, we added a morphology step before and changed the clone to auto threshold to b&w. before connected-components. We also moved the deskew to later and kept it in grayscale for some other processing steps we want to do before OCR.
